src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,281 @@
-        MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
-                               MPIDIG_context_id_to_comm(MPIDIG_REQUEST(rreq, context_id)),
-                               MPIDI_SHM_XPMEM_SEND_LMT_ACK, &ack_ctrl_hdr);
+        MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset, &seg_ptr, &src_buf);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    XPMEM_TRACE
+        ("handle_lmt_single_recv: handle matched rreq %p [source %d, tag %d, context_id 0x%x],"
+         " copy dst %p, src %lx, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
+         MPIDIG_REQUEST(rreq, tag), MPIDIG_REQUEST(rreq, context_id), (char *) MPIDIG_REQUEST(rreq,
+                                                                                              buffer),
+         src_offset, recv_data_sz);
+    /* Copy data to receive buffer */
+    MPIR_Localcopy(src_buf, recv_data_sz, MPI_BYTE,
+                   MPIDIG_REQUEST(rreq, buffer), MPIDIG_REQUEST(rreq, count), MPIDIG_REQUEST(rreq,
+                                                                                             datatype));
+
+    ack_ctrl_hdr.xpmem_slmt_send_fin.req_ptr = sreq_ptr;
+    mpi_errno = MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
+                                       MPIDIG_context_id_to_comm(MPIDIG_REQUEST
+                                                                 (rreq, context_id)),
+                                       MPIDI_SHM_XPMEM_SEND_LMT_SEND_FIN, &ack_ctrl_hdr);
     MPIR_ERR_CHECK(mpi_errno);
 
-    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
-    MPID_Request_complete(rreq);
+
+    ack_ctrl_hdr.xpmem_slmt_recv_fin.req_ptr = (uint64_t) rreq;
+    mpi_errno = MPIDI_XPMEM_ctrl_send_lmt_recv_fin_cb(&ack_ctrl_hdr);
+
+    mpi_errno = MPIDI_XPMEM_seg_deregist(seg_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
 
   fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_SINGLE_RECV);
     return mpi_errno;
   fn_fail:
     goto fn_exit;
 }
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_recv(uint64_t src_offset, size_t data_sz,
+                                                         uint64_t sreq_ptr, int src_lrank,
+                                                         MPIR_Comm * comm, MPIR_Request * rreq)
+{
+
+    int mpi_errno = MPI_SUCCESS;
+    int recvtype_iscontig;
+
+    /* TODO: need to support non-contig datatype on receiver side. */
+    MPIR_Datatype_is_contig(MPIDIG_REQUEST(rreq, datatype), &recvtype_iscontig);
+    if (recvtype_iscontig) {
+        /* Cooperative XPMEM copy */
+        mpi_errno =
+            MPIDI_XPMEM_handle_lmt_coop_recv(src_offset, data_sz, sreq_ptr, src_lrank, comm, rreq);
+    } else {
+        /* Receiver's data is non-contiguous, fall back to single-side XPMEM copy */
+        mpi_errno =
+            MPIDI_XPMEM_handle_lmt_single_recv(src_offset, data_sz, sreq_ptr, src_lrank, rreq);
+    }
+  fn_exit:
```

Comment:
Can we also fallback to single recv if `data_sz <= MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE`?

Reply:
You are right. I will fix it.